GetOldTweets3/manager/ConcurrentTweetManager.py

```diff
@@ -0,0 +1,78 @@
+# Author: Giulio Neusch-Frediani - www.github.com/giulionf
+from datetime import datetime
+from threading import Thread
+from queue import Queue
+from queue import Empty
+import copy
+
+from GetOldTweets3.manager.TweetManager import TweetManager
+
+DATE_FORMAT = "%Y-%m-%d"
+FIRST_TWEET_DATE = datetime.strptime("2006-03-21", DATE_FORMAT)
+
+
+class ConcurrentTweetManager:
+
+    @staticmethod
+    def getTweets(tweetCriteria, receiveBuffer=None, bufferLength=100, proxy=None, debug=False, worker_count=1, ordered=True):
+        # Init the queues
+        time_spans = Queue()
+        tweets = []
+        workers = []
+
+        # Split the date in smaller parts
+        since_date = datetime.strptime(tweetCriteria.since, DATE_FORMAT) if tweetCriteria.since else FIRST_TWEET_DATE
+        until_date = datetime.strptime(tweetCriteria.until, DATE_FORMAT) if tweetCriteria.until else datetime.now()
+        date_diff_per_worker = (until_date - since_date) / worker_count
+
+        # Create a TweetCriteria that can be cloned as Model by the Workers
+        criteria = copy.deepcopy(tweetCriteria)
+        criteria.setMaxTweets(criteria.maxTweets / worker_count)
+
+        # Create a time span for each of the splitted parts and a corresponding worker
+        for i in range(1, worker_count+1):
+            from_time = since_date + (i-1) * date_diff_per_worker
+            to_time = since_date + i * date_diff_per_worker
+            time_spans.put((from_time, to_time))
+
+            w = WorkerThread(copy.deepcopy(criteria), time_spans, tweets, receiveBuffer, bufferLength, proxy, debug)
+            w.start()
+            workers.append(w)
+
+        # Wait for the workers to finish and then stop them
+        time_spans.join()
+        for worker in workers:
+            worker.stop()
+
+        return sorted(tweets, key=lambda r: r.date)
```

Comment:
What do you think about creating temporary files like `.output_got_<some_unique_id>_<tweets_period>.csv` and concatenating them after all threads got finished?
I don't like the idea of storing all the tweets in memory during several hours of downloading.

The drawback of multithreaded downloading as that in case of an error the missing part will be in the middle of the date interval (in a single threaded downloading the missing part will always at the earliest date).

Reply:
I‘m not sure... In my opinion each application should implement that themselves with the receive_buffer callback if they really need it. What do you think?

For the second part: Yes thats true... But I can‘t think of a way to handle that issue... Any ideas?

Reply:
> For the second part: Yes thats true... But I can‘t think of a way to handle that issue... Any ideas?  

That was the reason why I don't want to implement parallel downloading. It makes things complex.

Reply:
> That was the first reason why I didn't want to implement parallel downloading. It complicates everything :) Not for developer, but for the user.

Thats certainly true, but it also has a lot of benefits :) I think seperating the classes will keep the best of both worlds! 

> I‘m not sure... In my opinion each application should implement that themselves with the receive_buffer callback if they really need it. What do you think?

What is your opinion on this idea?

> The second reason is that the user's IP will be banned by Twitter much faster

Does twitter really bann by IP? Is that not against their Privacy Policy?